libbpf-tools/bashreadline.c

```diff
@@ -0,0 +1,212 @@
+/* SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause) */
+/* Copyright (c) 2021 Facebook */
+#include <argp.h>
+#include <stdio.h>
+#include <errno.h>
+#include <signal.h>
+#include <time.h>
+#include <unistd.h>
+#include <ctype.h>
+
+#include <bpf/libbpf.h>
+#include <bpf/bpf.h>
+#include "bashreadline.h"
+#include "bashreadline.skel.h"
+#include "trace_helpers.h"
+#include "uprobe_helpers.h"
+
+#define PERF_BUFFER_PAGES	16
+#define PERF_POLL_TIMEOUT_MS	100
+#define warn(...) fprintf(stderr, __VA_ARGS__)
+
+static volatile sig_atomic_t exiting = 0;
+
+const char *argp_program_version = "bashreadline 1.0";
+const char *argp_program_bug_address =
+	"https://github.com/iovisor/bcc/tree/master/libbpf-tools";
+const char argp_program_doc[] =
+"Print entered bash commands from all running shells.\n"
+"\n"
+"USAGE: bashreadline [-s <path/to/libreadline.so>]\n"
+"\n"
+"EXAMPLES:\n"
+"    bashreadline\n"
+"    bashreadline -s /usr/lib/libreadline.so\n";
+
+static const struct argp_option opts[] = {
+	{ "shared", 's', "PATH", 0, "the location of libreadline.so library" },
+	{ NULL, 'h', NULL, OPTION_HIDDEN, "Show the full help" },
+	{},
+};
+
+static char *libreadline_path = NULL;
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+	switch (key) {
+	case 's':
+		libreadline_path = strdup(arg);
+		if (libreadline_path == NULL)
+			return ARGP_ERR_UNKNOWN;
+		break;
+	case 'h':
+		argp_state_help(state, stderr, ARGP_HELP_STD_HELP);
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+static void handle_event(void *ctx, int cpu, void *data, __u32 data_size)
+{
+	struct str_t *e = data;
+	struct tm *tm;
+	char ts[16];
+	time_t t;
+
+	time(&t);
+	tm = localtime(&t);
+	strftime(ts, sizeof(ts), "%H:%m:%S", tm);
+
+	printf("%-9s %-7d %s\n", ts, e->pid, e->str);
+}
+
+static void handle_lost_events(void *ctx, int cpu, __u64 lost_cnt)
+{
+	warn("lost %llu events on CPU #%d\n", lost_cnt, cpu);
+}
+
+static char *find_readline_so()
+{
+	const char *bash_path = "/bin/bash";
+	FILE *fp;
+	off_t func_off;
+	char *result = NULL;
+	char line[128];
+	int input_n;
+
+	func_off = get_elf_func_offset(bash_path, "readline");
+	if (func_off >= 0)
+		return strdup(bash_path);
+
+	/*
+	 * Try to find libreadline.so if readline is not defined in
+	 * bash itself.
+	 *
+	 * ldd will print a list of names of shared objects,
+	 * dependencies, and their paths.  The line for libreadline
+	 * would looks like
+	 *
+	 *      libreadline.so.8 => /usr/lib/libreadline.so.8 (0x00007b....)
+	 *
+	 * Here, it finds a line with libreadline.so and extracts the
+	 * path after the arrow, '=>', symbol.
+	 */
+	fp = popen("ldd /bin/bash", "r");
+	if (fp == NULL)
+		goto cleanup;
+	while ((input_n = fscanf(fp, "%*s => %127s", line)) != EOF) {
```

Comment:
even if this works, it will work by accident. You need to read a full line first, and then do sscanf(). I thought we already discussed this? See the following:

```
$ git diff
diff --git a/libbpf-tools/bashreadline.c b/libbpf-tools/bashreadline.c
index efe41bb5603c..d32a7180e227 100644
--- a/libbpf-tools/bashreadline.c
+++ b/libbpf-tools/bashreadline.c
@@ -87,6 +87,7 @@ static char *find_readline_so()
        int input_n;

        func_off = get_elf_func_offset(bash_path, "readline");
+       func_off = -1;
        if (func_off >= 0)
                return strdup(bash_path);

@@ -107,11 +108,12 @@ static char *find_readline_so()
        if (fp == NULL)
                goto cleanup;
        while ((input_n = fscanf(fp, "%*s => %127s", line)) != EOF) {
+               printf("line is '%s'\n", line);
                if (input_n <= 0)
                        continue;
                if (strstr(line, "/libreadline.so")) {
                        result = strdup(line);
-                       break;
+                       //break;
                }
        }


$ ldd /bin/bash
        linux-vdso.so.1 (0x00007ffe019d4000)
        libtinfo.so.6 => /lib64/libtinfo.so.6 (0x00007fc8cf7eb000)
        libdl.so.2 => /lib64/libdl.so.2 (0x00007fc8cf5e7000)
        libc.so.6 => /lib64/libc.so.6 (0x00007fc8cf222000)
        /lib64/ld-linux-x86-64.so.2 (0x00007fc8cfd42000)

$ sudo ./bashreadline
line is ''
line is ''
line is '/lib64/libtinfo.so.6'
line is '/lib64/libtinfo.so.6'
line is '/lib64/libdl.so.2'
line is '/lib64/libdl.so.2'
line is '/lib64/libc.so.6'
line is '/lib64/libc.so.6'
line is '/lib64/libc.so.6'
failed to find readline
```

Note the output of line. You should be getting only one instance of each library each time for this to work reliably, but clearly you are not. You fscanf() is just skipping at least one string from the input, but that's working by accident. You actually want to skip unrecognized line on each iteration, not parts of that line.

So please reimplement and validate it's working as expected.